### DISCUSSION TITLE
Help with HID library

### DISCUSSION THREAD

**Discussion by crecine** [micropython/micropython#13068] [Libraries & Drivers] [answered]
I apologize if this is the wrong place, but my question spans a couple of the categories, so I thought this was the most relevant.

I am building a keyboard with a Pi Pico and am trying to use the usb_hid library. I was originally using circuitpython, but some of my hardware uses micropython's machine library. The HID library in micropython doesn't seem to have all the features I need. I found that with circuitpython's Blinka, I am able to get usb_hid in micropython, but several dependencies are causing problems (atexit, typing, stat, fnmatch, etc). I would appreciate help either getting the libraries I need into my micropython lib or getting machine to work with circuitpython.

Thanks.

I am using micropython 1.21 and circuitpython 8.2.7.

**Comment by dhalbert** [comment]
(CircuitPython core developer here). What are you using `machine` for? I would steer you towards using just CircuitPython if you are using `usb_hid` and maybe `adafruit_hid`. If we reach that conclusion, we can move this support request to a CircuitPython area: https://adafru.io/discord or https://forums.adafruit.com/viewforum.php?f=60

**Comment by crecine** [comment]
Thanks for the quick response!
I'm not using it directly, but it is a dependency for some of the external hardware. It looks like it might only be using I2C and Pin, so I can probably hack together an alternative based on CircuitPython. It also uses time.ticks_ms. I'm pretty new to CircuitPython; is it possible to extend built-in libraries, like it is in [micropython](https://docs.micropython.org/en/latest/library/index.html#extending-built-in-libraries-from-python)?

The other thing I was interested in with micropython was precompiling for speed since key presses were sometimes missed when I was using CircuitPython. If that can be done with CircuitPython, then we can move this to the CircuitPython area as you suggested.

**Reply by dhalbert** [reply, marked as answer]
I2C and Pin of course have analogues in CircuitPython, so that should be easy. Note that `adafruit_bus_device` manages locking the I2C bus when it's in use. I suggest you look at a CircuitPython library that's similar to the I2C device you have. Or maybe there is already a CircuitPython library. If you're not sure, say what it is.

As for keypresses, have you seen `keypad`? https://learn.adafruit.com/key-pad-matrix-scanning-in-circuitpython

We have `supervisor.ticks_ms()` instead of `time.ticks_ms()`, and we also have the  `adafruit_ticks` library. Also note `keypad` includes timestamps for each key transition.

**Reply by crecine** [reply]
Thanks, it looks like that will work for me.
I hadn't seen keypad, I was just looping through the digital inputs.
